src/main/java/com/uid2/admin/vertx/service/SharingService.java

```diff
@@ -75,43 +75,81 @@ private void handleSetKeyset(RoutingContext rc) {
                 keysetProvider.loadContent();
                 siteProvider.loadContent();
             } catch (Exception e) {
-                LOGGER.error("Failed to load key acls");
+                LOGGER.error("Failed to load keysets");
                 rc.fail(500);
             }
 
+            final Map<Integer, Keyset> keysetsById = this.keysetProvider.getSnapshot().getAllKeysets();
+
             final JsonObject body = rc.body().asJsonObject();
 
             final JsonArray whitelist = body.getJsonArray("allowed_sites");
-            Integer keysetId = body.getInteger("keyset_id");
-            Integer siteId = body.getInteger("site_id");
-            String name = body.getString("name", "");
+            final Integer requestKeysetId = body.getInteger("keyset_id");
+            final Integer requestSiteId = body.getInteger("site_id");
+            final String requestName = body.getString("name", "");
 
-            if (keysetId == null && siteId == null
-                    || keysetId != null && siteId != null) {
+            if ((requestKeysetId == null && requestSiteId == null)
+                    || (requestKeysetId != null && requestSiteId != null)) {
                 ResponseUtil.error(rc, 400, "You must specify exactly one of: keyset_id, site_id");
                 return;
             }
 
-            final Map<Integer, Keyset> keysetsById = this.keysetProvider.getSnapshot().getAllKeysets();
-
-            if (keysetId != null) {
-                Keyset keyset = keysetsById.get(keysetId);
+            final int siteId;
+            final int keysetId;
+            final String name;
+            if(requestSiteId != null) {
+                siteId = requestSiteId;
+                name = requestName;
+                if (siteId == Const.Data.AdvertisingTokenSiteId
+                        || siteId == Const.Data.RefreshKeySiteId
+                        || siteId == Const.Data.MasterKeySiteId
+                        || siteProvider.getSite(siteId) == null
+                        || keysetsById.values().stream().anyMatch(k -> k.getSiteId() == siteId))  { // enforce single keyset per site
+                    ResponseUtil.error(rc, 400, "Site id " + siteId + " not valid");
+                    return;
+                }
+                keysetId = Collections.max(keysetsById.keySet()) + 1;
+                if (keysetsById.values().stream().anyMatch(item ->
+                        item.getSiteId() == siteId && item.getName().equalsIgnoreCase(name))) {
```

Comment:
We do duplicated 'site_id' check at line 107. this if-statement at 112 is redundant. keyset_name isn't unique.

Reply:
it will become relevant when we allow multiple keysets

Reply:
Added a comment here which references commented out test SharingServiceTest#KeysetSetNewIdenticalNameAndSiteId